### DISCUSSION TITLE
For Bazel String flags, is there a recommended way to use the string flag without discarding analysis cache?

### DISCUSSION THREAD

**Discussion by Ryang20718** [bazelbuild/bazel#22731] [Q&A]
For Bazel String flags, is there a recommended way to use the string flag without discarding analysis cache?

**Comment by fmeum** [comment]
@gregestren Is preserving the analysis cache across changes to Starlark flags only something you would consider doable? This could be an interesting sweet spot for improvements in this area.

**Reply by gregestren** [reply]
@susinmotion and @katre have been investigating this quarter the technical challenges of flag changes and analysis invalidation.

Tentative opinion is this isn't safe except for nodes we *know* aren't affected by the flag change.

That might make it possible to keep exec-configured nodes. Which can be a substantial part of the build graph. 

We should expect a more detailed summary soon.

@aranguyen has also been looking at limited flag propagation (i.e. strip out a flag when it goes outside the top-level project). Which might be able to fit in with this.

**Reply by gregestren** [reply]
Ideas here always welcome. I'd love to restrict the effect as much as we can do safely.

**Reply by fmeum** [reply]
Could we model Starlark flags as first-class Skyframe nodes rather than entries in a single `Map` so that change pruning could take care of this rather than us having to remember to drop the cache in certain cases? I can see how this would be a major undertaking, but are there any conceptual blockers to it?

**Reply by gregestren** [reply]
Interesting idea. I'm not totally sure it pans out but I like the thought. I'll give this some more thought.

**Reply by gregestren** [reply]
Starlark flags are of course also rules that consumers depend on to extract their values. I thought (but can't remember) that implied a Skyframe edge from each consuming configured target to the Starlark flags it reads. I think `select()` resolution would have similar effect. 

Unless I'm misremembering and we had distinct logic that executed the Starlark rule implementations and stored the results in a map.

If the above is true, I think we could adjust the logic to only invalidate Starlark flag rules themselves when their command-line values change. Which is much closer to optimal.

**Reply by gregestren** [reply]
For what it's worth, we have another initiative in the works related to [skyfocus](https://github.com/bazelbuild/bazel/commit/2fd449902cdc653f83267d5606bbc42f2f995dad) that might help this a lot.
